Mathlib/RingTheory/LocalRing/MaximalIdeal/Basic.lean

```diff
@@ -99,3 +106,33 @@ end LocalRing
 
 theorem LocalRing.maximalIdeal_eq_bot {R : Type*} [Field R] : LocalRing.maximalIdeal R = ⊥ :=
   LocalRing.isField_iff_maximalIdeal_eq.mp (Field.toIsField R)
+
+section Nilrad_max_localization
+
+open Ideal
+
+variable {R : Type*} [CommSemiring R] {S : Type*} [CommSemiring S] [Algebra R S] {M : Submonoid R}
+
+instance nilradmaxLocalRing (h : (nilradical R).IsMaximal) : LocalRing R := by
+  refine LocalRing.of_unique_max_ideal ⟨nilradical R, h, fun I hI ↦ ?_⟩
+  rw [nilradical_eq_sInf] at h ⊢
+  exact (IsMaximal.eq_of_le h hI.ne_top (sInf_le hI.isPrime)).symm
+
+/--
+Let `S` be the localization of a commutative semiring `R` at a submonoid `M` that does not
+contain 0. If the nilradical of `R` is maximal then there is a `R`-algebra isomorphism between
+`R` and `S`. -/
+noncomputable def nilradmaxlocalizationIsSelf (h : (nilradical R).IsMaximal) (h' : (0 : R) ∉ M)
+    [IsLocalization M S] : R ≃ₐ[R] S := by
+  have : LocalRing R := nilradmaxLocalRing h
```

Comment:
Since the condition for `LocalRing` can be automatically inferred, this line is unnecessary.

```suggestion

```